sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala

```diff
@@ -151,93 +162,146 @@ class SerializeConcatHostBuffersDeserializeBatch(
-      val (header, buffer) = SerializedHostTableUtils.readTableHeaderAndBuffer(in)
-      closeOnExcept(buffer) { _ =>
-        dataTypes = if (header.getNumColumns > 0) {
-          in.readObject().asInstanceOf[Array[DataType]]
-        } else {
-          Array.empty
+  /**
+   * Deserializes a broadcast result in the host into `data`, `numRows` and `dataLen`.
+   *
+   * Public for unit tests.
+   */
+  def doReadObject(in: ObjectInputStream): Unit = this.synchronized {
+    // no-op if we already have `batchInternal` or `data` set
+    if (batchInternal == null && data == null) {
+      withResource(new NvtxRange("DeserializeBatch", NvtxColor.PURPLE)) { _ =>
+        val (header, buffer) = SerializedHostTableUtils.readTableHeaderAndBuffer(in)
+        withResource(buffer) { _ =>
+          dataTypes = if (header.getNumColumns > 0) {
+            in.readObject().asInstanceOf[Array[DataType]]
+          } else {
+            Array.empty
+          }
+          // for a rowsOnly broadcast, null out the `data` member.
+          val rowsOnly = dataTypes.isEmpty
+          numRows = header.getNumRows
+          dataLen = header.getDataLen
+          data = if (!rowsOnly) {
+            JCudfSerialization.concatToHostBuffer(Array(header), Array(buffer))
+          } else {
+            null
+          }
         }
-        headers = Array(header)
-        buffers = Array(buffer)
       }
     }
   }
 
-  def numRows: Int = Option(batchInternal)
-    .map(_.numRows())
-    .getOrElse(headers.map(_.getNumRows).sum)
-
-  def dataSize: Long = Option(batchInternal)
-    .map(_.sizeInBytes)
-    .getOrElse(buffers.map(_.getLength).sum)
+  def dataSize: Long = dataLen
 
-  override def close(): Unit = this.synchronized {
-    buffers.safeClose()
-    buffers = Array.empty
-    Option(batchInternal).foreach(_.close())
+  /**
+   * This method is meant to only be called from `finalize` and it is not a regular
+   * AutoCloseable.close because we do not want Spark to close `batchInternal` when it spills
+   * the broadcast block's host torrent data.
+   *
+   * Reference: https://github.com/NVIDIA/spark-rapids/issues/8602
+   *
+   * Public for tests.
+   */
+  def closeInternal(): Unit = this.synchronized {
+    Seq(data, batchInternal).safeClose()
+    data = null
     batchInternal = null
   }
 
   @scala.annotation.nowarn("msg=method finalize in class Object is deprecated")
   override def finalize(): Unit = {
     super.finalize()
-    close()
+    closeInternal()
   }
 }
 // scalastyle:on no.finalize
 
+// scalastyle:off no.finalize
+/**
+ * Object used for executors to serialize a result for their partition that will be collected
+ * on the driver to be broadcasted out as part of the exchange.
+ * @param batch - GPU batch to be serialized and sent to the driver.
+ */
 @SerialVersionUID(100L)
 class SerializeBatchDeserializeHostBuffer(batch: ColumnarBatch)
   extends Serializable with AutoCloseable {
```

Comment:
This is still AutoCloseable -- intentional?

Reply:
This is AutoCloseable but it doesn't end up mattering since it's not the broadcast result. I can make it not AutoCloseable and instead implement closeInternal to make it clear...

Reply:
So I do end up using `.close` in this function https://github.com/NVIDIA/spark-rapids/pull/8617/files#diff-7ca82be5615724f3ba4376b82be1c8a34e3fb5df34a3dedfa9a21eb3e6e54b97R617, when creating a `SerializeConcatHostBuffersDeserializeBatch` from an array of `SerializeBatchDeserializeHostBuffer`